Implement `open`/`stat`/`mkdir`/`symlink` for standalone WASI mode

jiixyj opened the pull request:
Based on <https://github.com/emscripten-core/emscripten/pull/18285>, implement just enough syscalls to make the `test_fcntl_open` test from `test_core` work in standalone WASI mode.

Compared to #18285 I moved the preopen handling into their own `paths.{h,c}` component. The idea is that this component will also be responsible for cwd handling in a follow up PR.

Some Notes:

- The preopen handling is done before `main` runs. Therefore, I removed the locking functions.
- To enable the `test_fcntl_open` test I had to extend the `@also_with_standalone_wasm` decorator a bit. The test doesn't run on node and wasmer. Wasmer crashes when resolving symlinks. However, both wasmtime and toywasm successfully run the test.
- Standalone WASM engines now get a temporary `fs` folder inside the `out/test` folder to put temporary files in. This folder is cleaned after each WASM engine is done. I couldn't make this work for wasmer, but for wasmtime and toywasm it works very nicely.

jiixyj commented:
You can run the tests locally by adding some WASM engines to `WASM_ENGINES` in `.emscripten` and then:

```sh
./test/runner -v test_hello_argc_standalone
# or
./test/runner -v test_fcntl_open_standalone
```

jiixyj commented:
Oh, and I have patched wasmer a bit, so behavior compared to a release version may differ:

```patch
diff --git a/lib/wasix/src/syscalls/wasix/path_open2.rs b/lib/wasix/src/syscalls/wasix/path_open2.rs
index 6305b8bf59..c2297e05dc 100644
--- a/lib/wasix/src/syscalls/wasix/path_open2.rs
+++ b/lib/wasix/src/syscalls/wasix/path_open2.rs
@@ -148,7 +148,7 @@ pub(crate) fn path_open_internal(
     //
     // Maximum rights: should be the working dir rights
     // Minimum rights: whatever rights are provided
-    let adjusted_rights = /*fs_rights_base &*/ working_dir_rights_inheriting;
+    let adjusted_rights = fs_rights_base & working_dir_rights_inheriting;
     let mut open_options = state.fs_new_open_options();
 
     let target_rights = match maybe_inode {
@@ -276,7 +276,10 @@ pub(crate) fn path_open_internal(
                 }
             }
             Kind::Dir { .. } => {
-                if fs_rights_base.contains(Rights::FD_WRITE) {
+                if fs_rights_base.contains(Rights::FD_WRITE)
+                    || o_flags.contains(Oflags::TRUNC)
+                    || o_flags.contains(Oflags::CREATE)
+                {
                     return Ok(Err(Errno::Isdir));
                 }
             }
```

jiixyj commented:
The `other.test_standalone_whole_archive` test complains about missing `_fd_prestat_dir_name` and `_fd_prestat_get` symbols:

```
error: undefined symbol: fd_prestat_dir_name (referenced by root reference (e.g. compiled C/C++ code))                                                                                          
warning: To disable errors for undefined symbols use `-sERROR_ON_UNDEFINED_SYMBOLS=0`                                                                                                           
warning: _fd_prestat_dir_name may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: fd_prestat_get (referenced by root reference (e.g. compiled C/C++ code))                                                                                               
warning: _fd_prestat_get may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
```

Those are used in the preopen functions that run before main. Does anyone know how to fix this?

sbc100 commented:
fd_prestat_dir_name and fd_prestat_get are WASI syscalls which are implemented in libwasi.js.. seems strange that they would be reported as undefined symbols.

sbc100 commented on test/common.py:
```diff
@@ -1562,6 +1584,9 @@ def run_js(self, filename, engine=None, args=None,
         self.fail('JS subprocess failed (%s): %s (expected=%s).  Output:\n%s' % (error.cmd, error.returncode, assert_returncode, ret))
 
+    if run_in_tmpdir:
+      force_delete_contents(self.in_dir('fs'))
```
In general tests do not need to (indeed should not since helps with debugging) cleanup after themselves like this.

jiixyj replied:
That's a good point! While there already is some existing cleanup logic it only ran after each "full" test, not in between each WASM engine run when iterating over them. This made the test fail because each test expects a blank directory.

What do you think about a subdir per each engine? So `out/temp/wasmer`, `out/temp/wasmtime`, etc.? Then the existing cleanup code should work just as is.

jiixyj replied:
> What do you think about a subdir per each engine? So out/temp/wasmer, out/temp/wasmtime, etc.? Then the existing cleanup code should work just as is.

I implemented this and it works very well!

sbc100 commented on test/common.py:
```diff
@@ -1505,7 +1522,11 @@ def cleanup(line):
              interleaved_output=True,
-             input=None):
+             input=None,
+             run_in_tmpdir=False):
```
I wonder if we can avoid adding this extra logic to `run_js` which is use in a lot of different places.

Could we instead allow just add `cwd=None` here and allow higher level functions to pass CWD?

Is there some fundamental reason why the tests need to run in a subdirectory?

jiixyj replied:
> Good idea! I'll try to implement this.

Done!

sbc100 commented on test/common.py:
```diff
@@ -981,14 +997,14 @@ def get_nodejs(self):
     return config.NODE_JS_TEST
 
-  def require_node(self):
+  def require_node(self, allow_wasm_engines=False):
```
It seems odd to mark a test as `require_node` but also allow wasm engines.   Maybe we can just remove `require_node` from the tests in question instead?

jiixyj replied:
Yeah, I wonder about this, too. The original reason I did it like this comes from this line in `also_with_standalone_wasm`: https://github.com/emscripten-core/emscripten/blob/1ae691e9c61efe2d19fa2ab5e7cdd92f17596bf2/test/common.py#L639

Eventually this ends up in `require_engine()` where `self.wasm_engines` is set to an empty list. So what happened is that none of the WASM engines were actually used for any of the `also_with_standalone_wasm` tests!

Alternatively I could stash the `self.wasm_engines` in a local variable, then call `self.require_node()`, and then repopulate `self.wasm_engines`?

jiixyj replied:
> Alternatively I could stash the self.wasm_engines in a local variable, then call self.require_node(), and then repopulate self.wasm_engines?

I've implemented this, avoiding this `allow_wasm_engines` parameter.

jiixyj commented:
> fd_prestat_dir_name and fd_prestat_get are WASI syscalls which are implemented in libwasi.js.. seems strange that they would be reported as undefined symbols.

Ah, they were gated behind `PURE_WASI`! I could fix this by gating the new preopen functions in the .c code behind `EMSCRIPTEN_PURE_WASI` as well.

jiixyj commented:
To fix some failing CI tests I had to do some additional things:

- Updated wasmtime to v32.0.0
- Added stubs for some WASI syscalls to `libsigs.js`/`libwasi.js` and to scripts under `tools`

I also added attribution to wasi-libc to `LICENSE`. Does that look OK to you?

The `test-browser-chrome-2gb` failed in the latest run, but I think this may be just a bit flaky?

sbc100 commented on system/lib/standalone/paths.c:
```diff
@@ -0,0 +1,233 @@
+#include <string.h>
+#include <sysexits.h>
+
+/// A name and file descriptor pair.
```
How much of the code in this file is actually derived from the wasi-libc code?  If its actually derived from that codebase we should probably say so in the licence comment at the top.

Does wasi-libc use the triple-slash (///) for comments?   If not I would probably stick to normal C comments.

sbc100 commented on system/lib/standalone/paths.c:
```diff
@@ -0,0 +1,233 @@
+prefix_matches(const char* prefix, size_t prefix_len, const char* path) {
+  // Allow an empty string as a prefix of any relative path.
+  if (path[0] != '/' && prefix_len == 0)
+    return true;
```
IIRC out coding style uses braces around the body of single line `if` blocks (unless this is copied code in which case we should mention why the style is different in a comment up top.

jiixyj replied:
Yeah, this is wasi-libc derived code. I changed the brace style to match Emscripten's, as they aren't consistent about it anyway. Does this work for you?

sbc100 reviewed (commented):
Thanks for working on this!

I like the direction of this PR.  I've not yet had a chance for fully grok the code in paths.c/h but the rest is looking good

sbc100 commented on src/lib/libwasi.js:
```diff
@@ -608,6 +608,21 @@ var WasiLibrary = {
     return 0;
   },
+
+  path_filestat_get__nothrow: true,
```
Wrap these in `#if ALLOW_UNIMPLEMENTED_SYSCALLS` maybe?

jiixyj replied:
Sounds good! Let's see what the CI says...

sbc100 commented on test/common.py:
```diff
@@ -623,6 +623,9 @@ def metafunc(self, standalone):
       else:
+        nonlocal exclude_engines
+        if exclude_engines is None:
+          exclude_engines = []
```
I think you just put these two lines at the top of the function (before the `def decorated`) and avoid the nonlocal thing.

You could also just write `exclude_engines=[]` in the signature and add `# noqa` to silence the warning.

jiixyj replied:
done! `exclude_engines` isn't modified so doing it in the signature sounds good.

sbc100 commented on test/common.py:
```diff
@@ -635,9 +638,19 @@ def metafunc(self, standalone):
+          wasm_engines = []
+        else:
+          wasm_engines = [engine for engine in self.wasm_engines
+            if all(excluded not in os.path.basename(engine[0]) for excluded in exclude_engines)]
```
I find these double nested comprehension quite hard to understand.  Can we expand into a loop or split in to two statements perhaps?

How about:

```
wasm_engines = []
if not impure:
  for engine in self.wasm_engines:
     basename = os.path.basename(engine[0])
     if not any(pattern in basename for pattern in excluded_engines):
       wasm_engines.append(engine)
```

jiixyj replied:
Agreed, this is much more readable!

sbc100 commented on test/common.py:
```diff
@@ -614,7 +614,7 @@ def can_do_standalone(self, impure=False):
 # Impure means a test that cannot run in a wasm VM yet, as it is not 100%
 # standalone. We can still run them with the JS code though.
-def also_with_standalone_wasm(impure=False):
+def also_with_standalone_wasm(impure=False, exclude_engines=None):
```
Can you convert add some info about `exclude_engines` to the above comment?

Something like: "exclude_engines" is a list of engine names on which the test cannot run.  These are pattern that are matched against the basename of the engine executable.

sbc100 commented on test/common.py:
```diff
@@ -635,9 +638,19 @@ def metafunc(self, standalone):
+        else:
+          nodejs = self.require_node()
+          self.node_args += shared.node_bigint_flags(nodejs)
+        # `self.require_node()` clears `self.wasm_engines`, so set them here.
```
I don't love this, but we can refactor later I suppose.

sbc100 commented on test/common.py:
```diff
@@ -1966,11 +1981,15 @@ def _build_and_run(self, filename, expected_output, args=None,
+    for engine, subdir in engines:
+      if subdir:
+        subdir = self.in_dir(subdir)
+        ensure_dir(subdir)
```
So it looks for wasm engines only your want to run the JS in a new directory with the name of the engine?

I guess that will work for some tests, but don't some tests need to be run in alongside the files in the test directory?

Why is this needed? What happens if you don't do the chdir here?

jiixyj replied:
When I don't run inside a clean directory, the test for the second wasm engine would fail when trying to create `test-file` [here](<https://github.com/emscripten-core/emscripten/blob/f3d9eb42da8e7703fd319c9ca585b56ddad35732/test/fcntl/test_fcntl_open.c#L33>), as it already exists from the first wasm engine.

The current cleanup logic does not clean up the tests working directory after each engine, but only once in `setUp` for the whole thing.

> I guess that will work for some tests, but don't some tests need to be run in alongside the files in the test directory?

Yep, I have to admin that I don't have a good overview of the tests. But I thought that since the `wasm_engines` didn't have access to the filesystem before this PR, I at least can't break anything with this approach. The downside is of course that adding `@also_with_standalone_wasm` to a test like this won't work...

sbc100 replied:
I see.. so the problem only exists when running a single test with mutliple engines at the same time?

This is not something that gets a lot testing.. I for one basically never do this.  

I kind of wonder if we should remove this feature?   For now could we just run in the first wasm engine instead of in all of them?  

Actually the `EMTEST_USE_ALL_ENGINES` setting is false so enless you enable it we should only be running in a single engine.  Perhaps we should fix that?

sbc100 replied:
How about this solution:

- Leave `engines` as a simple list.
- Add a `self.directory_per_engine` test field which defaults to `false`
- Set this field to `true` in the tests that need it.
- Add this code here in the loop:

```
for engine, subdir in engines:
  cwd = None
  if self.directory_per_engine:
    # Some tests require a fresh directory for each engine, for example, because
    # they create file and are sensitive to existing files.
    dirname = os.path.basename(engine[0])
    ensure_dir(dirname)
    cwd = dirname
```

jiixyj replied:
Good idea, I'll try to implement this.

sbc100 commented on system/lib/standalone/standalone.c:
```diff
@@ -65,21 +104,106 @@ weak int _munmap_js(
+    } else {
+      return -EINVAL;
+    }
+  }
```
Is there some reason to add this extra scope here?

jiixyj replied:
Yes, it is there to limit `accmode`'s scope. I don't feel strongly about this, though.

sbc100 commented on system/lib/standalone/paths.c:
```diff
@@ -0,0 +1,254 @@
+oserr:
+  _Exit(EX_OSERR);
+software:
+  _Exit(EX_SOFTWARE);
```
Why not just `abort()` above instead of these goto + exit stuff?

jiixyj replied:
Good question -- this seems to be a wasi-libc thing. The code in `__main_void` (also derived from wasi-libc) uses `EX_OSERR` as well: <https://github.com/emscripten-core/emscripten/blob/f3d9eb42da8e7703fd319c9ca585b56ddad35732/system/lib/standalone/__main_void.c#L36>

sbc100 commented:
This change looks like its almost ready to land.

Before we do though I wanted to check what your motivation for trying push this forward is?   Is there some use case where you want to use emscripten to target non-web engines?  Is there some reason why you need to use emscripten and not just wasi-sdk?

jiixyj commented:
> Before we do though I wanted to check what your motivation for trying push this forward is? Is there some use case where you want to use emscripten to target non-web engines? Is there some reason why you need to use emscripten and not just wasi-sdk?

For the fun of it I'm trying to port a hobby project of mine (CLI tool written in C++, using exceptions for error handling) to WASI. I actually tried wasi-sdk a few months ago but couldn't quite get exceptions to work. Then a few weeks ago I saw that Emscripten supported both a `STANDALONE_WASM`/`PURE_WASI` mode and even out of the box support for the new "exnref" exceptions (using `-sWASM_LEGACY_EXCEPTIONS=0`). So in the end it was faster for me to get to a working `.wasm` blob using Emscripten. But yeah, "on paper", wasi-sdk is more suitable for my use case...

sbc100 commented on system/lib/standalone/paths.c:
```diff
@@ -0,0 +1,254 @@
+    const preopen* pre = &preopens[i];
+    assert(pre->prefix != NULL);
+    assert(pre->fd != (__wasi_fd_t)-1);
+#ifdef __wasm__
```
Why `ifdef __wasm__` here?     IIUC this code only compiles and run on wasm

sbc100 commented on system/lib/standalone/paths.c:
```diff
@@ -0,0 +1,254 @@
+  preopen_capacity = new_capacity;
+  free(old_preopens);
+
+  assert_invariants();
```
Why not just `preopens = realloc(preopen, new_capacity)` here?

sbc100 commented on system/lib/standalone/paths.c:
```diff
@@ -0,0 +1,254 @@
+      path++;
+    } else if (path[0] == '.' && path[1] == '/') {
+      path += 2;
+    } else if (path[0] == '.' && path[1] == 0) {
```
use '`\0`' here for the terminator?

This case looks like its trying to match `.` on its own and then return the empty string.. is that right?

sbc100 commented on system/lib/standalone/paths.c:
```diff
@@ -0,0 +1,254 @@
+        if (!register_preopened_fd(fd, prefix)) {
+          goto software;
+        }
+        free(prefix);
```
Given that `register_preopened_fd` is going to store and use the prefix, why not pass owership?  i.e. don't call `free()` here or `strdup` in `register_preopened_fd`?

kripken reviewed (commented):
(otherwise no comments aside from @sbc100 's)
